### PR TITLE
compress: support streaming lz4 without full input mmap

### PR DESCRIPTION
**Pull request by bluca**
The advantage of stream compression is keeping a low memory profile,
but the lz4 stream compressor usage mmaps the whole file in memory.

Change it to read bits by bits, like the other stream compression
helpers.

This will come in handy when moving systemd-coredump to do compression first, then decompression, to avoid using large amount of disk space in /var/lib/coredump. I'll follow-up with that later, but this requisite is stand-alone as the mmap-less implementation looks good to have regardless.

### CODE REVIEW

File: src/libsystemd/sd-journal/compress.c [hunk 1]
```diff
@@ -630,35 +629,35 @@ int compress_stream_lz4(int fdf, int fdt, uint64_t max_bytes) {
...
         frame_size = LZ4F_compressBound(LZ4_BUFSIZE, &preferences);
-        size =  frame_size + 64*1024; /* add some space for header and trailer */
-        buf = malloc(size);
-        if (!buf)
+        out_allocsize = frame_size + 64*1024; /* add some space for header and trailer */
+        out_buff = malloc(out_allocsize);
+        if (!out_buff)
+                return -ENOMEM;
+
+        in_allocsize = frame_size; // TODO: what's the optimal size?
```
Comment (bluca) [inline src/libsystemd/sd-journal/compress.c]: Not a compression expert, so looking for suggestion here: what's the best way to size the input buffer? Should it be some % of the output one?

Comment (bluca) [reply src/libsystemd/sd-journal/compress.c]: nvm we compress at most LZ4_BUFSIZE anyway

**Comment by bluca** [comment]
The test fails under clang 11+ and sanitizers, where the compressed+decompressed file doesn't match the original. The sizes are identical. Still trying to figure it out, there must be something silly I've missed.

```
/* test compression */
Buffer size is 589836 bytes, header size 7 bytes.
LZ4 compression finished (1282520 -> 617533 bytes, 48.1%)
Binary files /tmp/sys/buildclang/test-compress and - differ
Assertion 'system(cmd) == 0' failed at src/libsystemd/sd-journal/test-compress.c:199, function void test_compress_stream(const char *, const char *, compress_stream_t *, decompress_stream_t *, const char *)(). Aborting.
```

**Comment by bluca** [comment]
> The test fails under clang 11+ and sanitizers, where the compressed+decompressed file doesn't match the original. The sizes are identical. Still trying to figure it out, there must be something silly I've missed.
> 
> ```
> /* test compression */
> Buffer size is 589836 bytes, header size 7 bytes.
> LZ4 compression finished (1282520 -> 617533 bytes, 48.1%)
> Binary files /tmp/sys/buildclang/test-compress and - differ
> Assertion 'system(cmd) == 0' failed at src/libsystemd/sd-journal/test-compress.c:199, function void test_compress_stream(const char *, const char *, compress_stream_t *, decompress_stream_t *, const char *)(). Aborting.
> ```

weird, it now always passed - triggered 3/4 times

**Comment by bluca** [comment]
> > The test fails under clang 11+ and sanitizers, where the compressed+decompressed file doesn't match the original. The sizes are identical. Still trying to figure it out, there must be something silly I've missed.
> > ```
> > /* test compression */
> > Buffer size is 589836 bytes, header size 7 bytes.
> > LZ4 compression finished (1282520 -> 617533 bytes, 48.1%)
> > Binary files /tmp/sys/buildclang/test-compress and - differ
> > Assertion 'system(cmd) == 0' failed at src/libsystemd/sd-journal/test-compress.c:199, function void test_compress_stream(const char *, const char *, compress_stream_t *, decompress_stream_t *, const char *)(). Aborting.
> > ```
> 
> weird, it now always passed - triggered 3/4 times

So this failure is only reproducible with clang 11.0.1, but not with the newer 11.1.0. Only when enabling ASAN/UBSAN and -O1. I am inclined to think this is an issue in the asan library, triggered by the change (somehow).

To reproduce, in a Debian unstable/testing build environment (which has clang 11.0.1):

CC=clang CXX=clang++ meson buildclang -Dfuzz-tests=true -Db_lundef=false -Db_sanitize=address,undefined --optimization=1 -Dtests=unsafe

and then run test-compress

**Comment by mrc0mmand** [comment]
Interestingly enough it works when you link the ASan library dynamically (which gcc does by default), i.e.:

```
# CFLAGS=-shared-libasan CXXFLAGS=-shared-libasan CC=clang-11 CXX=clang++-11 meson build-clang11-shared -Db_lundef=false -Db_sanitize=address,undefined --optimization=1
# ninja -C build-clang11-shared
# LD_LIBRARY_PATH=/usr/lib/llvm-11/lib/clang/11.0.1/lib/linux/ ASAN_OPTIONS=verbosity=5:print_stats=1 build-clang11-shared/test-compress
...
/* testing LZ4 compression */
/* create source from build-clang11-shared/test-compress */
/* test compression */
Buffer size is 589836 bytes, header size 7 bytes.
LZ4 compression finished (78328 -> 36231 bytes, 46.3%)
/* test decompression */
LZ4 decompression finished (36231 -> 78328 bytes, 216.2%)
/* test faulty decompression */
LZ4 decompression finished (36231 -> 78328 bytes, 216.2%)
Decompressed stream longer than 78327 bytes
/* test_lz4_decompress_partial */
...
/* test_decompress_startswith_short with ZSTD */
==38124==Attached to thread 38108.
==38124==Detached from thread 38108.
```

When linking statically (done by default by clang), the issue reappears:
```
# CC=clang-11 CXX=clang++-11 meson build-clang11-static -Db_lundef=false -Db_sanitize=address,undefined --optimization=1
# ninja -C build-clang11-static
# ASAN_OPTIONS=verbosity=5:print_stats=1 build-clang11-static/test-compress
/* testing LZ4 compression */
/* create source from build-clang11-static/test-compress */
/* test compression */
Buffer size is 589836 bytes, header size 7 bytes.
LZ4 compression finished (1282528 -> 617564 bytes, 48.2%)
Binary files build-clang11-static/test-compress and - differ
Assertion 'system(cmd) == 0' failed at src/libsystemd/sd-journal/test-compress.c:199, function void test_compress_stream(const char *, const char *, compress_stream_t *, decompress_stream_t *, const char *)(). Aborting.
Aborted (core dumped)
```

However, if you give `test-compress` a different binary (which isn't instrumented), everything works even with the static ASan version:
```
# ASAN_OPTIONS=verbosity=5:print_stats=1 build-clang11-static/test-compress /usr/bin/clang-11
...
/* test compression */
Buffer size is 589836 bytes, header size 7 bytes.
LZ4 compression finished (128696 -> 71532 bytes, 55.6%)
/* test decompression */
LZ4 decompression finished (71532 -> 128696 bytes, 179.9%)
/* test faulty decompression */
LZ4 decompression finished (71532 -> 128696 bytes, 179.9%)
Decompressed stream longer than 128695 bytes
/* test_lz4_decompress_partial */
...
/* test_decompress_startswith_short with ZSTD */
==42606==Attached to thread 42590.
==42606==Detached from thread 42590.
```

**Comment by mrc0mmand** [comment]
Ugh, if you use the "static" `test-compress` binary as an input to the "dynamic" one, it doesn't work either:

```
# LD_LIBRARY_PATH=/usr/lib/llvm-11/lib/clang/11.0.1/lib/linux/ ASAN_OPTIONS=verbosity=5:print_stats=1 build-clang11-shared/test-compress build-clang11-static/test-compress
...
/* testing LZ4 compression */
/* create source from build-clang11-static/test-compress */
/* test compression */
Buffer size is 589836 bytes, header size 7 bytes.
LZ4 compression finished (1282528 -> 617564 bytes, 48.2%)
Binary files build-clang11-static/test-compress and - differ
Assertion 'system(cmd) == 0' failed at src/libsystemd/sd-journal/test-compress.c:199, function void test_compress_stream(const char *, const char *, compress_stream_t *, decompress_stream_t *, const char *)(). Aborting.
Aborted (core dumped)
```

**Comment by mrc0mmand** [comment]
So, it turns out it happens even with unsanitized and unoptimized clang/gcc builds, when using the clang-sanitized `test-compress` binary as an input:

```
# meson build-gcc
# ninja -C build-gcc
# build-gcc/test-compress build-clang11-static/test-compress
...
/* testing LZ4 compression */
/* create source from build-clang11-static/test-compress */
/* test compression */
Buffer size is 589836 bytes, header size 7 bytes.
LZ4 compression finished (1282528 -> 617564 bytes, 48.2%)
Binary files build-clang11-static/test-compress and - differ
Assertion 'system(cmd) == 0' failed at src/libsystemd/sd-journal/test-compress.c:199, function test_compress_stream(). Aborting.
Aborted (core dumped)
```

Attaching the "offending" `test-compress` binary (unpack it before using): [test-compress.tar.gz](https://github.com/systemd/systemd/files/6125598/test-compress.tar.gz)

**Comment by bluca** [comment]
So! It turns out, this was caused by the ```options.stableSrc = 1``` flag - removing it, the issue doesn't reproduce anymore.
The best I can figure out is that it governs some internal optimization here:

https://github.com/lz4/lz4/blob/9061a0a93582854cbddb64cb048ae527a0c220af/lib/lz4frame.c#L895

which was fine to use when the source buffer was a single mmapped file in memory, but it is no longer fine to use with a small circular buffer. Why and how this manifests in libasan getting mangled, I could not possibly tell, and for my own sanity would rather never find out.

PTAL!

**Comment by poettering** [comment]
lgtm

**Pull request merged**